File: uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/list/views/CheckmarkButtonView.java
/*
 * Copyright (C) 2016 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.activities.habits.list.views;

import android.content.*;
import android.content.res.*;
import android.graphics.*;
import android.support.annotation.*;
import android.text.*;
import android.util.*;
import android.view.*;

import org.isoron.androidbase.utils.*;
import org.isoron.uhabits.*;
import org.isoron.uhabits.activities.*;
import org.isoron.uhabits.core.preferences.*;
import org.isoron.uhabits.utils.*;

import static android.view.View.MeasureSpec.EXACTLY;
import static android.view.View.MeasureSpec.makeMeasureSpec;
import static org.isoron.androidbase.utils.InterfaceUtils.getDimension;
import static org.isoron.androidbase.utils.InterfaceUtils.getFontAwesome;
import static org.isoron.uhabits.core.models.Checkmark.CHECKED_EXPLICITLY;
import static org.isoron.uhabits.core.models.Checkmark.UNCHECKED;
import static org.isoron.uhabits.utils.AttributeSetUtils.getIntAttribute;

public class CheckmarkButtonView extends View
{
    private int color;

    private int value;

    private StyledResources styledRes;

    private TextPaint paint;

    private int lowContrastColor;

    private RectF rect;

    @Nullable
    private Preferences prefs;

    @NonNull
    private OnToggleListener onToggleListener;

    @NonNull
    private OnInvalidToggleListener onInvalidToggleListener;

    public CheckmarkButtonView(@Nullable Context context)
    {
        super(context);
        init();
    }

    public CheckmarkButtonView(@Nullable Context ctx, @Nullable AttributeSet attrs)
    {
        super(ctx, attrs);
        init();

        if(ctx == null) throw new IllegalStateException();
        if(attrs == null) throw new IllegalStateException();

        int paletteColor = getIntAttribute(ctx, attrs, "color", 0);
        setColor(PaletteUtils.getAndroidTestColor(paletteColor));
        int value = getIntAttribute(ctx, attrs, "value", 0);
        setValue(value);
    }

    public void setColor(int color)
    {
        this.color = color;
        postInvalidate();
    }

    public void setValue(int value)
    {
        this.value = value;
        postInvalidate();
    }

    public void performToggle()
    {
        onToggleListener.onToggle();
        value = (value == CHECKED_EXPLICITLY ? UNCHECKED : CHECKED_EXPLICITLY);
        performHapticFeedback(HapticFeedbackConstants.LONG_PRESS);
        postInvalidate();
    }

    @Override
    protected void onDraw(Canvas canvas)
    {
        super.onDraw(canvas);
        Resources resources = getResources();

        paint.setColor(value == CHECKED_EXPLICITLY ? color : lowContrastColor);
        int id = (value == UNCHECKED ? R.string.fa_times : R.string.fa_check);
        String label = resources.getString(id);
        float em = paint.measureText("m");

        rect.set(0, 0, getWidth(), getHeight());
        rect.offset(0, 0.4f * em);
        canvas.drawText(label, rect.centerX(), rect.centerY(), paint);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec)
    {
        Resources res = getResources();
        int height = res.getDimensionPixelSize(R.dimen.checkmarkHeight);
        int width = res.getDimensionPixelSize(R.dimen.checkmarkWidth);

        widthMeasureSpec = makeMeasureSpec(width, EXACTLY);
        heightMeasureSpec = makeMeasureSpec(height, EXACTLY);

        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    private void init()
    {
        setFocusable(false);

        styledRes = new StyledResources(getContext());

        paint = new TextPaint();
        paint.setTypeface(getFontAwesome(getContext()));
        paint.setAntiAlias(true);
        paint.setTextAlign(Paint.Align.CENTER);
        paint.setTextSize(getDimension(getContext(), R.dimen.smallTextSize));

        rect = new RectF();
        color = Color.BLACK;
        lowContrastColor = styledRes.getColor(R.attr.lowContrastTextColor);

        onToggleListener = () -> {};
        onInvalidToggleListener = () -> {};

        if (getContext() instanceof HabitsActivity)
        {
            HabitsApplicationComponent component =
                ((HabitsActivity) getContext()).getAppComponent();
            prefs = component.getPreferences();
        }

        setOnClickListener((v) -> {
            if (prefs == null) return;
            if (prefs.isShortToggleEnabled()) performToggle();
            else onInvalidToggleListener.onInvalidToggle();
        });

        setOnLongClickListener(v -> {
            performToggle();
            return true;
        });
    }

    public void setOnInvalidToggleListener(
        @NonNull OnInvalidToggleListener onInvalidToggleListener)
    {
        this.onInvalidToggleListener = onInvalidToggleListener;
    }

    public void setOnToggleListener(@NonNull OnToggleListener onToggleListener)
    {
        this.onToggleListener = onToggleListener;
    }

    public interface OnInvalidToggleListener
    {
        void onInvalidToggle();
    }

    public interface OnToggleListener
    {
        void onToggle();
    }
}

File: uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/list/views/CheckmarkButtonView.kt
/*
 * Copyright (C) 2016-2021 Álinson Santos Xavier <git@axavier.org>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.activities.habits.list.views

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.text.TextPaint
import android.view.HapticFeedbackConstants
import android.view.View
import android.view.View.MeasureSpec.EXACTLY
import org.isoron.platform.gui.ScreenLocation
import org.isoron.uhabits.R
import org.isoron.uhabits.core.models.Entry
import org.isoron.uhabits.core.models.Entry.Companion.NO
import org.isoron.uhabits.core.models.Entry.Companion.SKIP
import org.isoron.uhabits.core.models.Entry.Companion.UNKNOWN
import org.isoron.uhabits.core.models.Entry.Companion.YES_AUTO
import org.isoron.uhabits.core.models.Entry.Companion.YES_MANUAL
import org.isoron.uhabits.core.preferences.Preferences
import org.isoron.uhabits.inject.ActivityContext
import org.isoron.uhabits.utils.drawNotesIndicator
import org.isoron.uhabits.utils.getCenter
import org.isoron.uhabits.utils.getFontAwesome
import org.isoron.uhabits.utils.sp
import org.isoron.uhabits.utils.sres
import org.isoron.uhabits.utils.toMeasureSpec
import javax.inject.Inject

const val TOGGLE_DELAY_MILLIS = 2000L

class CheckmarkButtonViewFactory
@Inject constructor(
    @ActivityContext val context: Context,
    val preferences: Preferences
) {
    fun create() = CheckmarkButtonView(context, preferences)
}

class CheckmarkButtonView(
    context: Context,
    val preferences: Preferences
) : View(context),
    View.OnClickListener,
    View.OnLongClickListener {

    var color: Int = Color.BLACK
        set(value) {
            field = value
            invalidate()
        }

    var value: Int = 0
        set(value) {
            field = value
            invalidate()
        }

    var notes = ""
        set(value) {
            field = value
            invalidate()
        }

    var onToggle: (Int, String, Long) -> Unit = { _, _, _ -> }

    var onEdit: (ScreenLocation) -> Unit = { _ -> }

    private var drawer = Drawer()

    init {
        setOnClickListener(this)
        setOnLongClickListener(this)
    }

    fun performToggle(delay: Long) {
        value = Entry.nextToggleValue(
            value = value,
            isSkipEnabled = preferences.isSkipEnabled,
            areQuestionMarksEnabled = preferences.areQuestionMarksEnabled
        )
        onToggle(value, notes, delay)
        performHapticFeedback(HapticFeedbackConstants.LONG_PRESS)
        invalidate()
    }

    override fun onClick(v: View) {
        if (preferences.isShortToggleEnabled) performToggle(TOGGLE_DELAY_MILLIS)
        else onEdit(getCenter())
    }

    override fun onLongClick(v: View): Boolean {
        if (preferences.isShortToggleEnabled) onEdit(getCenter())
        else performToggle(TOGGLE_DELAY_MILLIS)
        return true
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        drawer.draw(canvas)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val height = resources.getDimensionPixelSize(R.dimen.checkmarkHeight)
        val width = resources.getDimensionPixelSize(R.dimen.checkmarkWidth)
        super.onMeasure(
            width.toMeasureSpec(EXACTLY),
            height.toMeasureSpec(EXACTLY)
        )
    }

    private inner class Drawer {
        private val rect = RectF()
        private val bgColor = sres.getColor(R.attr.cardBgColor)
        private val lowContrastColor = sres.getColor(R.attr.contrast40)
        private val mediumContrastColor = sres.getColor(R.attr.contrast60)

        private val paint = TextPaint().apply {
            typeface = getFontAwesome()
            isAntiAlias = true
            textAlign = Paint.Align.CENTER
        }

        fun draw(canvas: Canvas) {
            paint.color = when (value) {
                YES_MANUAL, YES_AUTO, SKIP -> color
                NO -> {
                    if (preferences.areQuestionMarksEnabled) mediumContrastColor
                    else lowContrastColor
                }
                else -> lowContrastColor
            }
            val id = when (value) {
                SKIP -> R.string.fa_skipped
                NO -> R.string.fa_times
                UNKNOWN -> {
                    if (preferences.areQuestionMarksEnabled) R.string.fa_question
                    else R.string.fa_times
                }
                else -> R.string.fa_check
            }
            paint.textSize = when {
                id == R.string.fa_question -> sp(12.0f)
                value == YES_AUTO -> sp(13.0f)
                else -> sp(14.0f)
            }
            if (value == YES_AUTO) {
                paint.strokeWidth = 5f
                paint.style = Paint.Style.STROKE
            } else {
                paint.strokeWidth = 0f
                paint.style = Paint.Style.FILL
            }

            val label = resources.getString(id)
            val em = paint.measureText("m")

            rect.set(0f, 0f, width.toFloat(), height.toFloat())
            rect.offset(0f, 0.4f * em)
            canvas.drawText(label, rect.centerX(), rect.centerY(), paint)

            if (value == YES_AUTO) {
                paint.color = bgColor
                paint.style = Paint.Style.FILL
                canvas.drawText(label, rect.centerX(), rect.centerY(), paint)
            }

            drawNotesIndicator(canvas, color, em, notes)
        }
    }
}
